### ISSUE TITLE
multiple calls to getopt using new options.c screws up the argument vector

### ISSUE THREAD

**Issue by flihp** [01org/tpm2.0-tools#126]
As part of the expected behavior for getopt, "unknown" arguments (those not in the options structure array / argument string) get moved to the end of the argument vector. This is getopt trying to be helpful by moving what it things are positional parameters to the end of the vector for the calling application. This backfires spectacularly if you call getopt(_long)? twice looking for different options each time.

By way of example suppose we have a tool `tpm2_foo` that expects a single option `-c|--capability` with a string argument. So `-c bar` would be an acceptable argument to pass to it. If we also specifiy a TCTI by name using the `[-T|--tcti]=my-tcti` option we may end up executing the command like: `tpm2_foo -c some-arg -T my-tcti`. Since the common options are processed first the option processing in the `tpm2_foo` program will happen after the common options are processed.

The issue however arises in this initial option processing since the `-c` option is unknown to the common option processing function. The current implementation will end up transforming the argv from:
```
argv[0]: tpm2_foo
argv[1]: -c
argv[2]: some-arg
argv[3]: -T
argv[4]: my-tcti
```
to
```
argv[0]: tpm2_foo
argv[1]: -c
argv[2]: -T
argv[3]: my-tcti
argv[4]: some-arg
```
This causes the second around of option processing with getopt in the tool for tool specific options to get all messed up.

**Comment by flihp** [comment]
The easiest approach to resolving this would be to prefix the optstring with `-` to prevent getopt from rewritng the argv. This solves the immediate problem but it still leaves the common options in the argument vector so the tool specific option processing code will have to ignore them. At first glance this doesn't look too bad but if the command is working with positional parameters it will have to use the `-` trick as well and force the caller to provide all positional parameters after `-` style options (not the end of the world, but limiting).

**Comment by flihp** [comment]
The ideal solution would be for the getopt loop in the common options.c file to parse *and remove* all of the common options from the argument vector. This means that the option processing code for each tool would be able to do whatever they want with their own options.

This however will require some pretty complex code since the common option processing will have to do significantly more than it does currently.

**Comment by williamcroberts** [comment]
What about using argparse, will that behave differently? https://www.gnu.org/software/libc/manual/html_node/Parsing-Program-Arguments.html

I am assuming at this point we can just consider GLIBC as our target environment.

**Comment by williamcroberts** [comment]
Also, I am looking at how you implemented this, you may want to have a const char *tool_get_options(), where the tool can return its option string for option parsing.

**Comment by flihp** [comment]
I'm a big fan of `arg_parse`. But in the interest of minimizing the changes required for each tool I stuck with `getopt_long` (for better or for worse). If we do want to move to `arg_parse` in the future I wouldn't object but I'd like to get the issues with the current `getopt` implementation sorted first.

**Comment by williamcroberts** [comment]
I think getopt() will actually make it harder for you since this re-ordering is built in. Also, I think it will make it harder in the future as well to add new support. The current interface lacks a lot of support, that argparse provides, like argument grouping.
I would switch to argparse now, and have the following interfaces that tools can implement:
/* provide their options */
struct argp_option *tool_get_options();

/* can build a static data structure with their option values
* and get a chance to attempt conversions to bool, integer, custom, etc.
 * true means all is well, false on parse error.
 */
bool tool_on_option(char key, const char *value)

/* execute tool passing leftover arguments */
int tool_main(int argc, char *argv);

If you want I can provide you these source files to drop in.

**Comment by flihp** [comment]
Would love to see the implementation you've got. Plz send.

**Comment by williamcroberts** [comment]
[example.zip](https://github.com/01org/tpm2.0-tools/files/670026/example.zip)

**Comment by williamcroberts** [comment]
argp supports children parsers, that made this "much" simpler. The help output and such is very nice too:

$ ./options --help
Usage: options [OPTION...] stuff and junk tools stuff and junk
Tool Common

 Tool XXX [OPTIONS] [STRINGS...]
  -f, --foo                  print foo status
  -r, --repeat[=COUNT]       Repeat the output COUNT (default 10) times

 Common Options:
  -o, --output=FILE          Output to FILE instead of standard output
  -v, --verbose              Produce verbose output

 This is my custom tool

  -?, --help                 Give this help list
      --usage                Give a short usage message
  -V, --version              Print program version

Mandatory or optional arguments to long options are also mandatory or optional
for any corresponding short options.

**Comment by flihp** [comment]
If you can create a topic branch and post a link to your branch it'll be a lot easier to experiment with.

I've always been a fan of the built in `--help`stuff with argparse. But the current implementation invokes `man` when the `--help` argument is provided so it's less compelling.

**Comment by williamcroberts** [comment]
It's not a full branch, it's a sample code of how to do it. main.c should be fairly complete. You would need to add all the common options you wish to handle to main.c and then update the tools to use tool.h interface. I would drop options.c and options.h.

Integration into the tooling is left as a separate exercise. What happens if man is not available, for instance if these ever were to get ported to Android?

If you want help to pull up a man page, I think you can use: ARGP_NO_HELP in flags per: https://www.gnu.org/software/libc/manual/html_node/Argp-Flags.html#Argp-Flags
and then create your own help handler. You could even do --help --help=man so on man you do the exec and on --help you just call arp_help().

**Comment by williamcroberts** [comment]
I could probably create a branch for this next week since you're off early next week.

**Comment by flihp** [comment]
"off" is a relative term. Mostly it just means I won't be checking my work email. I'll be online and working on OSS stuff and I'll still be camping out on IRC etc.

I'm about to create a PR with my first existing tool port to the new stuff. We can figure out how / when to work in the argparse stuff in parallel.

I wish there was a better way to comment on stuff "inline" like an email here but I don't really know how to here. But anyways the bit about --help / executing man and Android: for "embeddedish" stuff like android I think it's safe to assume that there won't be anyone at the keyboard (if there is one) and so not having manuals wasn't an issue. For those doing development the manuals can be obtained elsewhere.

**Comment by williamcroberts** [comment]
Based on discussion, we'll port all the tools to dynamic TCTI before handling this correctly by switching all the tools to an argparse approach.

**Comment by flihp** [comment]
resolved in 5e23adb62d957c2292909e7de7684f8db3026a68
